types/time.go

```diff
@@ -562,165 +564,444 @@ func ParseDateFormat(format string) []string {
 	format = strings.TrimSpace(format)
 
 	start := 0
-	var seps = make([]string, 0)
-	for i := 0; i < len(format); i++ {
-		// Date format must start and end with number.
-		if i == 0 || i == len(format)-1 {
-			if !unicode.IsNumber(rune(format[i])) {
-				return nil
-			}
-
-			continue
+	var seps = make([]string, 0, maxSQLTimeParts)
+	if len(format) == 0 {
+		return nil
+	}
+	// Date format must start with number.
+	if !unicode.IsNumber(rune(format[0])) {
+		return nil
+	}
+	for i := 1; i < len(format); i++ {
+		if len(seps) > maxSQLTimeParts { //max len is 7,'Y-M-D H-M-S.ms'
+			break
 		}
 
-		// Separator is a single none-number char.
-		if !unicode.IsNumber(rune(format[i])) {
-			if !unicode.IsNumber(rune(format[i-1])) {
-				return nil
-			}
+		var isNumberIB1 = unicode.IsNumber(rune(format[i-1]))
+		var isNumberI = unicode.IsNumber(rune(format[i]))
 
+		//if i-1 is not number,i is number,meaning the beginning of one sep
+		if !isNumberIB1 && isNumberI {
+			start = i
+		}
+
+		//if i - 1 is number,i is not number,meaning the ending of one sep
+		if isNumberIB1 && !isNumberI {
 			seps = append(seps, format[start:i])
-			start = i + 1
+			start = -1
```

Comment:
I think `start = i + 1` is more accurate.

Reply:
It can not set  `start = i + 1`.
Thinking of `Select cast('1701020302.11BB' as datetime);` , the last if logic in ParseDateFormat can not be executed:
`
if 0 <= start && start < len(format){
  seps = append(seps, format[start:])
}
`